### ISSUE TITLE
Error with Function Breakpoints: Debug adapter process has terminated unexpectedly

### ISSUE THREAD

**Issue by jonathanmedd** [PowerShell/vscode-powershell#1159]
### System Details

- Operating system name and version: Windows 10, 10.0.16299
- VS Code version: 1.19.1
- PowerShell extension version: 1.5.1
- Output from `$PSVersionTable`:

```
Major  Minor  Build  Revision
-----  -----  -----  --------
1      5      1      0

DotJoshJohnson.xml@1.9.2
eamodio.gitlens@7.2.0
ms-vscode.PowerShell@1.5.1


Name                           Value
----                           -----
PSVersion                      6.0.0-rc.2
PSEdition                      Core
GitCommitId                    v6.0.0-rc.2
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


Name                           Value
----                           -----
PSVersion                      5.1.16299.98
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.16299.98
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

```

### Issue Description

I am experiencing a problem with setting a function breakpoint and using the VS Code debugger. The issue occurs in both PS Core 6 RC2 and PS 5.1.

Example code:

```
1..10 | ForEach-Object {Write-Host "Hello World"}
```

Create a function breakpoint for Write-Host:

![image](https://user-images.githubusercontent.com/4677545/34731794-48651cd8-f55b-11e7-9598-07494def6d81.png)

will produce the following error on PS Core 6 when starting the debugger. (PS 5.1 does not produce error text)

An error has occurred that was not properly handled. Additional information is shown below. The PowerShell process will exit.

Unhandled Exception: 1..10 | ForEach-Object {Write-Host "Hello World"}
System.ArgumentException: Unexpected breakpoint type: CommandBreakpoint
   at Microsoft.PowerShell.EditorServices.BreakpointDetails.Create(Breakpoint breakpoint)
   at Microsoft.PowerShell.EditorServices.Protocol.Server.DebugAdapter.<DebugService_BreakpointUpdated>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

Hit Command breakpoint on 'Write-Host'

PowerShell Core 6 RC2

![image](https://user-images.githubusercontent.com/4677545/34731319-e7b33c90-f559-11e7-9f4b-5dd125f64b15.png)

PowerShell 5.1

![image](https://user-images.githubusercontent.com/4677545/34731540-86323ca4-f55a-11e7-9f0e-b056bfd300fa.png)

### Attached Logs

[DebugLogs.zip](https://github.com/PowerShell/vscode-powershell/files/1615894/DebugLogs.zip)

**Comment by rkeithhill** [comment]
Can you save the file to disk and try again.  Right now, setting breakpoints won't work for "in-memory" files.  That said, we should fix the PSES crash.

**Comment by jonathanmedd** [comment]
Same problem after saving to disk. I should have mentioned that in the original description, although I think the logs were from a non-saved.

**Comment by rkeithhill** [comment]
Doh!  :-)

**Comment by jonathanmedd** [comment]
I can reproduce the error with any saved file. For example opening DebugTest.ps1 which ships as an example as part of the VS Code PowerShell extension. Create a function breakpoint for Write-Host and it will blow up with the same error:

![image](https://user-images.githubusercontent.com/4677545/34748365-0ea65366-f594-11e7-8974-8a935f6e2b30.png)

Here are the logs for this one: 
[DebugLogs2.zip](https://github.com/PowerShell/vscode-powershell/files/1617007/DebugLogs2.zip)

**Comment by tylerl0706** [comment]
Thanks for this report, @jonathanmedd!

@rkeithhill: I didn't see too much in the logs but that exception he supplied seems like a good start, don't you think?

**Comment by SeeminglyScience** [comment]
@tylerl0706 looks like this line

https://github.com/PowerShell/PowerShellEditorServices/blob/33cb42140d23ac94c28d325d242899b9a39aae7e/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs#L1038

needs to be updated to check for `CommandBreakpoint` and `VariableBreakpoint` and create either `CommandBreakpointDetails` or `VariableBreakpointDetails`

**Comment by realonedet** [comment]
I've been having the exact same issue (on windows and mac), posted previously with no resolution, has this been root caused yet ?

**Comment by rkeithhill** [comment]
I've got a fix in the works for this.  In the process I found another issue - I think with VSCode's debug protocol.  There is a BreakpointEvent that allows the debug adapter to notify VSCode about `line` breakpoints changes that occur in the debug adapter.  I see no equivalent for function breakpoint changes.

I could be missing something in the debug protocol spec but just in case, I've raised an issue with the VSCode folks.
